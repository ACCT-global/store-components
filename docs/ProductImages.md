ðŸ“¢ Don't fork this project. Use, [contribute](https://github.com/vtex-apps/awesome-io#contributing), or open issues through [Store Discussion](https://github.com/vtex-apps/store-discussion).

# Product Images

`ProductImages` is a VTEX component that render a set of Image or Video of a product.
This Component can be imported and used by any VTEX app.

![image](https://user-images.githubusercontent.com/284515/70234551-a5c8cc00-173f-11ea-87d9-9f95c79761c8.png)

## Configuration

1. Import the vtex.store-component's app to your theme's dependencies in the manifest.json, for example:

```json
  dependencies: {
    "vtex.store-components": "3.x"
  }
```

2. Add the `product-images` block to any block bellow `store.product`. For example:

```json
  "store.product": {
    "children": [
      "flex-layout.row#product",
    ]
  },
  "flex-layout.row#product": {
    "children": [
      "product-images"
    ]
  },
  "product-images": {
    "props": {
      "displayThumbnailsArrows": true
    }
  },
```

| Prop name                 | Type      | Description                                                                                                 | Default Value |
| ------------------------- | --------- | ----------------------------------------------------------------------------------------------------------- | ------------- |
| `thumbnailsOrientation`   | `Enum`    | Choose the orientation of the thumbnails. Set to `vertical` or `horizontal`                                 | `vertical`    |
| `position`                | `Enum`    | Set the position of the thumbnails(`left` or `right`). Only used when `thumbnailsOrientation` is `vertical` | `left`        |
| `displayThumbnailsArrows` | `boolean` | Displays navigation arrows on the thumbnails if there are enough thumbnails for them to scroll              | `false`       |
| `hiddenImages`       | `string`  | Hides images whose labels match the values listed in this prop. Intended to be used along with the `product-summary-sku-selector` block. You can have more information at the [SKUSelector docs](/docs/SKUSelector.md) | `skuvariation` |
| `aspectRatio`             | `string`                                   | Sets the aspect ratio of the image; that is, whether the image should be square, portrait, landscape, etc. The value should follow the [common aspect ratio notation](https://en.wikipedia.org/wiki/Aspect_ratio_(image)), i.e., two numbers separated by a colon (e.g, "1:1" for square, "3:4" for upright portrait, or even large values such as "1920:1080") | `"auto"`          |
| `showNavigationArrows`             | `boolean`                                   | Controls if the navigation arrows should appear | `true`          |
| `showPaginationDots`             | `boolean`                                   | Controls if the pagination dots should appear | `true`          |
| `thumbnailAspectRatio`             | `string`                                   | Sets the aspect ratio of the thumbnail image; For more information about aspect ratio, check the `aspectRatio` prop | `"auto"`          |
| `thumbnailMaxHeight`             | `number`                                   | The max height for the thumbnail image | `true`          |
| `zoomMode`                | `disabled\|in-place-click\|in-place-hover` | Sets the zoom behavior.                                                                                                                                                                                                                                                                                                                                         | `in-place-click` |
| `zoomFactor`              | `number`                                   | Sets how much the zoom increases the image size (e.g. `2` will make the zoomed-in image twice as large)                                                                                                                                                                                                                                                              | 2                |

## Customization

In order to apply CSS customizations in this and other blocks, follow the instructions given in the recipe on [Using CSS Handles for store customization](https://vtex.io/docs/recipes/style/using-css-handles-for-store-customization).

| CSS Handles |
| --- |
| `productImagesContainer` |
| `video`                  |
| `image`                  |
| `carouselCursorDefault`   |
| `carouselInconCaretRight` |
| `carouselIconCaretLeft`   |
| `carouselGaleryThumbs`    |
| `carouselThumbBorder`     |
| `carouselGaleryCursor`    |
| `carouselImagePlaceholder`   |
