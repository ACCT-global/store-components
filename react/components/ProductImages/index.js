import PropTypes from 'prop-types'
import React, { Component } from 'react'

import Carousel from './components/Carousel'


class ProductImages extends Component {
  render() {
    const {images} = this.props

    if(images.length === 0) return null

    const slides = images.map(image => {
      return {
        type: 'image',
        urls: image.imageUrls,
        thresholds: image.thresholds || [],
        thumbUrl: image.thumbnailUrl || image.imageUrls[0],
        alt: image.imageText,
      }})

    return (
      <div className="w-100">
        <Carousel slides={slides}/>
      </div>)
  }
}

ProductImages.propTypes = {
  /** Array of images to be passed for the Thumbnail Slider component as a props */
  images: PropTypes.arrayOf(
    PropTypes.shape({
      /** URL of the image */
      imageUrls: PropTypes.arrayOf(PropTypes.string.isRequired),
      /** Size thresholds used to choose each image */
      thresholds: PropTypes.arrayOf(PropTypes.number),
      /** URL of the image thumbnail */
      thumbnailUrl: PropTypes.string,
      /** Text that describes the image */
      imageText: PropTypes.string.isRequired,
    })
  ),
}

ProductImages.defaultProps = {
  images: []
}

export default ProductImages
