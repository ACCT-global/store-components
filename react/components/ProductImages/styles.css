/**
*   Product Image CSS Module
*/

.content {
}

.video {
}

.image {
}

.imageBlur30 {
  content: ‘’;
  background: inherit;
  box-shadow: inset 0 0 0 200px rgba(255, 255, 255, 0.3);
  filter: blur(10px);
}

.imageTransitionOpacity {
  transition: 1s;
}

.carouselGaleryCursor {
  cursor: url('./components/Carousel/cursor.svg') 8 8, default;
}

.carouselGaleryThumbs :global(.swiper-slide) {
  height: auto;
}

.carouselGaleryThumbs
  :global(.swiper-slide):not(:global(.swiper-slide-thumb-active))
  .carouselThumbBorder {
  border-width: 0px;
}

.carouselIconCaretRight,
.carouselIconCaretLeft {
  stroke: white;
  stroke-width: 0.5px;
  -webkit-filter: drop-shadow(0px 1px 1px #cacbcc);
  filter: drop-shadow(0px 1px 1px #cacbcc);
}

.carouselCursorDefault:hover {
  cursor: default;
}

.slider {
  position: absolute;
  height: 5px;
  overflow-x: hidden;
}

.subline {
  position: absolute;
  height: 5px;
}
.inc {
  animation: increase 2s infinite;
}
.dec {
  animation: decrease 2s 0.5s infinite;
}

@keyframes increase {
  from {
    left: -5%;
    width: 5%;
  }
  to {
    left: 130%;
    width: 100%;
  }
}
@keyframes decrease {
  from {
    left: -80%;
    width: 80%;
  }
  to {
    left: 110%;
    width: 10%;
  }
}
