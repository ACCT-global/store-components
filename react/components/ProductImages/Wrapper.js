import * as React from 'react'
import { ProductContext } from 'vtex.product-context'
import { path, isEmpty } from 'ramda'

import ProductImages from './index'
import { changeImageUrlSize } from './utils/generateUrl'
import { 
  getThumbnailsPositionNames, 
  getThumbnailsPositionValues, 
  thumbnailsPosition ,
} from './utils/thumbnailPositionEnum'

const thresholds = [640]
const imageSizes = [1280, 1920]
const thumbnailSize = 160

const ProductImagesWrapper = ({ thumbnailPosition, ...props}) => {
  const valuesFromContext = React.useContext(ProductContext)

  const productImagesProps = () => {
    if (!valuesFromContext || isEmpty(valuesFromContext)) 
      return {
        ...props,
        position: thumbnailPosition,
      }

    return {
      images: getImages(),
      position: thumbnailPosition,
    }
  }

  const getImages = () => {
    const { selectedItem } = valuesFromContext

    const images = path(['images'], selectedItem)

    if (!images) return []
    return images.map(image => ({
      imageUrls: imageSizes.map(size =>
        changeImageUrlSize(image.imageUrl, size)
      ),
      thresholds,
      thumbnailUrl: changeImageUrlSize(image.imageUrl, thumbnailSize),
      imageText: image.imageText,
    }))
  }

  return (
    <ProductImages { ...productImagesProps() } />
  )
}

ProductImagesWrapper.schema = {
  // @TODO review title and description
  title: 'editor.product-details.title',
  description: 'editor.product-details.description',
  type: 'object',
  properties: {
    thumbnailPosition: {
      title: 'editor.product-details.thumbnailsPosition.title',
      type: 'string',
      enum: getThumbnailsPositionValues(),
      enumNames: getThumbnailsPositionNames(),
      default: thumbnailsPosition.DISPLAY_LEFT.value,
      isLayout: false,    
    },
  },
}

export default ProductImagesWrapper