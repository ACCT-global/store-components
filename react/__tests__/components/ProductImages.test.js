import React from 'react'
import { render } from 'enzyme'

import ProductImages from '../../ProductImages'

describe('<ProductImages /> component', () => {
  function renderComponent(customProps = {}) {
    const component = <ProductImages {...customProps} />

    return render(component)
  }

  it('should be able to mount and not break', () => {
    const component = renderComponent()
    expect(component).toBeTruthy()
  })

  it('should match snapshot without images', () => {
    const component = renderComponent()
    expect(component).toMatchSnapshot()
  })

  it('should match snapshot with images', () => {
    const component = renderComponent({ images: [] })
    expect(component).toMatchSnapshot()
  })
})
