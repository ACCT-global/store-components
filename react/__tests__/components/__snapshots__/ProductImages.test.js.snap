// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImages /> should match the snapshot with images 1`] = `
<DocumentFragment>
  <div
    class="content w-100"
  >
    <div
      class="w-100 aspect-ratio aspect-ratio--1x1"
    >
      <svg
        class="content-loader-mock"
      >
        <rect
          height="100%"
          width="100%"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ProductImages /> should match the snapshot with no images 1`] = `
<DocumentFragment>
  <div
    class="content w-100"
  >
    <svg
      fill="none"
      height="100%"
      viewBox="0 0 512 512"
      width="100%"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="#F2F2F2"
        height="512"
        width="512"
      />
      <rect
        height="150.474"
        stroke="#CACBCC"
        stroke-width="2"
        width="144.286"
        x="183.857"
        y="180.2"
      />
      <path
        d="M183.78 303.688H328.214"
        stroke="#CACBCC"
        stroke-width="2"
      />
      <path
        d="M205.082 279.563L223.599 240.507L242.116 260.035L269.892 220.979L306.926 279.563H205.082Z"
        stroke="#CACBCC"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
      <path
        d="M252.225 213.939C252.225 219.822 247.66 224.52 242.114 224.52C236.569 224.52 232.004 219.822 232.004 213.939C232.004 208.057 236.569 203.359 242.114 203.359C247.66 203.359 252.225 208.057 252.225 213.939Z"
        stroke="#CACBCC"
        stroke-width="2"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`<ProductImages /> should match the snapshot with thumbnails in right position 1`] = `
<DocumentFragment>
  <div
    class="content w-100"
  >
    <svg
      fill="none"
      height="100%"
      viewBox="0 0 512 512"
      width="100%"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="#F2F2F2"
        height="512"
        width="512"
      />
      <rect
        height="150.474"
        stroke="#CACBCC"
        stroke-width="2"
        width="144.286"
        x="183.857"
        y="180.2"
      />
      <path
        d="M183.78 303.688H328.214"
        stroke="#CACBCC"
        stroke-width="2"
      />
      <path
        d="M205.082 279.563L223.599 240.507L242.116 260.035L269.892 220.979L306.926 279.563H205.082Z"
        stroke="#CACBCC"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
      <path
        d="M252.225 213.939C252.225 219.822 247.66 224.52 242.114 224.52C236.569 224.52 232.004 219.822 232.004 213.939C232.004 208.057 236.569 203.359 242.114 203.359C247.66 203.359 252.225 208.057 252.225 213.939Z"
        stroke="#CACBCC"
        stroke-width="2"
      />
    </svg>
  </div>
</DocumentFragment>
`;
